File: workers/plpcg-catalog/src/auth/username_rules.ts
/** Handle público: 3–30 chars, a-z 0-9 _, sempre em minúsculas. */
const USERNAME_RE = /^[a-z0-9_]{3,30}$/;

export function normalizeUsername(raw: string): string {
  return raw.trim().toLowerCase();
}

/** Retorna mensagem de erro ou null se válido. */
export function validateUsername(raw: unknown): string | null {
  if (typeof raw !== 'string') return 'username required';
  const normalized = normalizeUsername(raw);
  if (normalized.length === 0) return 'username required';
  if (!USERNAME_RE.test(normalized)) {
    return 'invalid username';
  }
  return null;
}

File: lib/features/auth/domain/username_rules.dart
/// Espelha `workers/.../username_rules.ts` — handle 3–30: a-z 0-9 _.
abstract final class UsernameRules {
  static final RegExp pattern = RegExp(r'^[a-z0-9_]{3,30}$');

  static String normalize(String raw) => raw.trim().toLowerCase();

  /// Null se válido; senão chave de erro estável.
  static String? validate(String raw) {
    final normalized = normalize(raw);
    if (normalized.isEmpty) return 'username required';
    if (!pattern.hasMatch(normalized)) return 'invalid username';
    return null;
  }
}
